#!/bin/sh -e
## 05_handle-short-read.dpatch
##
## DP: Description: Handle read(2) returning less data than we asked for.
## DP: Author: Steve Langasek <vorlon@debian.org>
## DP: Upstream status: Not submitted
## DP: Date: 2004-10-17

if [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch}"

case "$1" in
       -patch) patch $patch_opts -p1 < $0;;
       -unpatch) patch $patch_opts -p1 -R < $0;;
        *)
                echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
                exit 1;;
esac

exit 0

@DPATCH@
diff -urNad --exclude=CVS --exclude=.svn ./bucket.c /tmp/dpep-work.eiim0G/gdbm-1.8.3/bucket.c
--- ./bucket.c	1999-05-19 01:16:05.000000000 +0100
+++ /tmp/dpep-work.eiim0G/gdbm-1.8.3/bucket.c	2006-04-24 03:18:01.000000000 +0100
@@ -31,7 +31,7 @@
 #include "autoconf.h"
 
 #include "gdbmdefs.h"
-
+#include <errno.h>
 
 /* Initializing a new hash buckets sets all bucket entries to -1 hash value. */
 void
@@ -68,7 +68,8 @@
      int dir_index;
 {
   off_t bucket_adr;	/* The address of the correct hash bucket.  */
-  int   num_bytes;	/* The number of bytes read. */
+  int   num_bytes = 0; /* The total number of bytes read. */
+  int   bytes_read;    /* Number of bytes read in this syscall */
   off_t	file_pos;	/* The return address for lseek. */
   register int index;	/* Loop index. */
 
@@ -111,7 +112,12 @@
       if (file_pos != bucket_adr)
 	_gdbm_fatal (dbf, "lseek error");
 
-      num_bytes = read (dbf->desc, dbf->bucket, dbf->header->bucket_size);
+      do
+        {
+          bytes_read = read (dbf->desc, dbf->bucket+num_bytes, dbf->header->bucket_size-num_bytes);
+          num_bytes += bytes_read;
+        }
+      while ((bytes_read > 0 || errno == EINTR) && num_bytes < dbf->header->bucket_size);
       if (num_bytes != dbf->header->bucket_size)
 	_gdbm_fatal (dbf, "read error");
     }
diff -urNad --exclude=CVS --exclude=.svn ./falloc.c /tmp/dpep-work.eiim0G/gdbm-1.8.3/falloc.c
--- ./falloc.c	2006-04-24 03:17:54.000000000 +0100
+++ /tmp/dpep-work.eiim0G/gdbm-1.8.3/falloc.c	2006-04-24 03:18:01.000000000 +0100
@@ -31,7 +31,7 @@
 #include "autoconf.h"
 
 #include "gdbmdefs.h"
-
+#include <errno.h>
 
 /* The forward definitions for this file.  See the functions for
    the definition of the function. */
@@ -174,7 +174,8 @@
 pop_avail_block (dbf)
      gdbm_file_info *dbf;
 {
-  int  num_bytes;		/* For use with the read system call. */
+  int  num_bytes = 0;          /* For use with the read system call. */
+  int  bytes_read;             /* For use with the read system call. */
   off_t file_pos;		/* For use with the lseek system call. */
   avail_elem new_el;
   avail_block *new_blk;
@@ -199,7 +200,12 @@
   /* Read the block. */
   file_pos = lseek (dbf->desc, new_el.av_adr, L_SET);
   if (file_pos != new_el.av_adr)  _gdbm_fatal (dbf, "lseek error");
-  num_bytes = read (dbf->desc, new_blk, new_el.av_size);
+  do
+    {
+      bytes_read = read (dbf->desc, new_blk+num_bytes, new_el.av_size-num_bytes);
+      num_bytes += bytes_read;
+    }
+  while ((bytes_read > 0 || errno == EINTR) && num_bytes < new_el.av_size);
   if (num_bytes != new_el.av_size) _gdbm_fatal (dbf, "read error");
 
   /* Add the elements from the new block to the header. */
diff -urNad --exclude=CVS --exclude=.svn ./findkey.c /tmp/dpep-work.eiim0G/gdbm-1.8.3/findkey.c
--- ./findkey.c	1999-05-19 01:16:06.000000000 +0100
+++ /tmp/dpep-work.eiim0G/gdbm-1.8.3/findkey.c	2006-04-24 03:18:01.000000000 +0100
@@ -31,6 +31,7 @@
 #include "autoconf.h"
 
 #include "gdbmdefs.h"
+#include <errno.h>
 
 
 /* Read the data found in bucket entry ELEM_LOC in file DBF and
@@ -41,11 +42,12 @@
      gdbm_file_info *dbf;
      int elem_loc;
 {
-  int num_bytes;		/* For seeking and reading. */
+  int num_bytes = 0;           /* For seeking and reading. */
   int key_size;
   int data_size;
   off_t file_pos;
   data_cache_elem *data_ca;
+  int bytes_read;
 
   /* Is it already in the cache? */
   if (dbf->cache_entry->ca_data.elem_loc == elem_loc)
@@ -74,7 +76,12 @@
 		    dbf->bucket->h_table[elem_loc].data_pointer, L_SET);
   if (file_pos != dbf->bucket->h_table[elem_loc].data_pointer)
     _gdbm_fatal (dbf, "lseek error");
-  num_bytes = read (dbf->desc, data_ca->dptr, key_size+data_size);
+  do
+    {
+      bytes_read = read (dbf->desc, data_ca->dptr+num_bytes, key_size+data_size-num_bytes);
+      num_bytes += bytes_read;
+    }
+  while ((bytes_read > 0 || errno == EINTR) && num_bytes < key_size+data_size);
   if (num_bytes != key_size+data_size) _gdbm_fatal (dbf, "read error");
   
   return data_ca->dptr;
